### PR TITLE
Generate init module

### LINKED ISSUE #511: Create an initialization module for each program
Currently, we create an LLVM module for each Ballerina module. With this issue, we will create an additional initialization module.

This module will be responsible for defining the program's main function, which will have a fixed name like `_bal_main`, and is called by C `main` function. This will call the user-defined main function in the root module, whose name depends on the name of root module (call this the "user main"). This will allow us to fix #456.

The initialization module will expand to deal with other things (which will be separate issues):
 * initialize every module in the program #512
 * process command-line arguments, turning the OS-provided arguments into the form expected by the user main
 * deal with possibility that LLVM return type of user main may be void or a tagged ptr
 * provide data structures the the program modules need for performing runtime type-checking #402 
 
For a root module `foo.bal` we will generate the initialization module into `foo._init.ll`. (We should implement #444 to ensure this cannot clash with a user module name.) The testing Makefile needs to deal with this.

nback will export a function that builds an initialization module, which will be called by compile.bal. The parameters to that will include
- a list of the module ids of all the programs modules, in the order in which the modules should be initialized (or the reverse of that)
- the ModuleExports for the root module (so that we know how to call main)

This is a prerequisite for #495.

### PR DESCRIPTION
**Pull request by manuranga**
Fixes #511

### CODE REVIEW

File: compiler/modules/nback/basic.bal [hunk 1]
```diff
@@ -222,7 +222,8 @@ function buildFunctionDecl(Scaffold scaffold, bir:ExternalSymbol symbol, bir:Fun
     }
 }
 
-function buildFunctionSignature(bir:FunctionSignature signature, err:Location loc) returns llvm:FunctionType|BuildError {
+// pr-todo: remove loc param
+function buildFunctionSignature(bir:FunctionSignature signature, err:Location? loc = ()) returns llvm:FunctionType|BuildError {
```
Comment (manuranga) [inline compiler/modules/nback/basic.bal]: I am planing to remove  `err:Location loc`.
@heshanpadmasiri

File: compiler/modules/nback/init.bal [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    check buildMain(entryModId, userMainName, publicFuncs.get(userMainName), llMod, builder);
+    return llMod;
+}
+
+public function buildMain(bir:ModuleId entryModId, string userMainName, bir:FunctionSignature userMainSig, llvm:Module llMod, llvm:Builder builder) returns BuildError? {
+    llvm:FunctionType ty = { returnType: "void", paramTypes: [] };
+    llvm:FunctionDefn func = llMod.addFunctionDefn("_bal_main", ty);
+    builder.positionAtEnd(func.appendBasicBlock());
+
+    string userMainMangledName = mangleInternalSymbol(entryModId, {identifier: userMainName, isPublic: true});
```
Comment (jclark) [inline compiler/modules/nback/init.bal]: Mapping constructor missing spaces

File: compiler/modules/nback/init.bal [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import wso2/nballerina.bir;
+import wso2/nballerina.print.llvm;
+
+public function buildInitModule(bir:ModuleId entryModId, map<bir:FunctionSignature> publicFuncs, llvm:Context llContext) returns llvm:Module|BuildError {
+    llvm:Module llMod = llContext.createModule();
+    llvm:Builder builder = llContext.createBuilder();
+    // SUBSET support cases with no/different entry function
+    string userMainName = "main";
+    check buildMain(entryModId, userMainName, publicFuncs.get(userMainName), llMod, builder);
```
Comment (jclark) [inline compiler/modules/nback/init.bal]: front-end isn't requiring main: it just requires that if main exists, then it has to have a particular signature and be public. Not sure we want to require main for E test cases.  A Ballerina program doesn't have to have a main (think something that just starts a server). So I think for now our _bal_main should just not call anything if there is no user main. Maybe we can add an error/warning down the road.

File: compiler/modules/nback/basic.bal (hunk 1 above)
Comment (jclark) [reply compiler/modules/nback/basic.bal]: Move this into build.bal since it's going to be used by init.bal.

File: compiler/compile.bal [hunk 4]
```diff
@@ -68,7 +72,9 @@ class CompileContext {
 function compileBalFile(string filename, string basename, string? outputBasename, nback:Options nbackOptions, OutputOptions outOptions) returns CompileError? {
     CompileContext cx = new(basename, outputBasename, nbackOptions, outOptions);
     front:ResolvedModule mod = check processModule(cx, DEFAULT_ROOT_MODULE_ID, [ {filename} ], cx.outputFilename());
+    // pr-todo: find if this should be moved to init.bal
```
Comment (jclark) [inline compiler/compile.bal]: No, because nback doesn't depend on front.

**Review by jclark — CHANGES_REQUESTED** [review]
Does sub.mk need changes or does it work as is?

File: compiler/compile.bal (hunk 4 above)
Comment (manuranga) [reply compiler/compile.bal]: I didn't mean just the method call, I mean move the whole logic to `init.bal`.

File: runtime/main.c [hunk 5]
```diff
@@ -2,12 +2,12 @@
 #include "hash.h"
 
 char *_bal_stack_guard;
-extern void BAL_ROOT_NAME(main)();
```
Comment (manuranga) [inline runtime/main.c:5]: This is no longer used, but I didn't remove the definition.

File: compiler/modules/nback/init.bal (hunk 2 above)
Comment (manuranga) [reply compiler/modules/nback/init.bal]: Ahg fixed, only if we had a style guide.

**Comment by manuranga** [comment]
It works as is, didn't need to change `sub.mk`.

File: compiler/compile.bal (hunk 4 above)
Comment (manuranga) [reply compiler/compile.bal]: Discussed in the meeting, this logic belongs in the frontend.

File: compiler/modules/nback/init.bal [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+public function buildInitModule(bir:ModuleId entryModId, map<bir:FunctionSignature> publicFuncs, llvm:Context llContext) returns llvm:Module|BuildError {
+    llvm:Module llMod = llContext.createModule();
+    llvm:Builder builder = llContext.createBuilder();
+    string userMainName = "main";
+    check buildMain(entryModId, userMainName, publicFuncs[userMainName], llMod, builder);
+    return llMod;
+}
+
+public function buildMain(bir:ModuleId entryModId, string userMainName, bir:FunctionSignature? userMainSig, llvm:Module llMod, llvm:Builder builder) returns BuildError? {
```
Comment (jclark) [inline compiler/modules/nback/init.bal]: Shouldn't be public.

File: compiler/modules/nback/init.bal [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import wso2/nballerina.bir;
+import wso2/nballerina.print.llvm;
+
+public function buildInitModule(bir:ModuleId entryModId, map<bir:FunctionSignature> publicFuncs, llvm:Context llContext) returns llvm:Module|BuildError {
+    llvm:Module llMod = llContext.createModule();
+    llvm:Builder builder = llContext.createBuilder();
+    string userMainName = "main";
```
Comment (jclark) [inline compiler/modules/nback/init.bal]: I would make this a const for now.

File: compiler/modules/nback/init.bal [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+public function buildMain(bir:ModuleId entryModId, string userMainName, bir:FunctionSignature? userMainSig, llvm:Module llMod, llvm:Builder builder) returns BuildError? {
+    llvm:FunctionType ty = { returnType: "void", paramTypes: [] };
+    llvm:FunctionDefn func = llMod.addFunctionDefn("_bal_main", ty);
+    builder.positionAtEnd(func.appendBasicBlock());
+
+    if userMainSig != () {
+        string userMainMangledName = mangleInternalSymbol(entryModId, { identifier: userMainName, isPublic: true });
+        llvm:FunctionType userMainTy = check buildFunctionSignature(userMainSig);
+        llvm:FunctionDecl userMainDecl = llMod.addFunctionDecl(userMainMangledName, userMainTy);
+        llvm:Value? retValue = builder.call(userMainDecl, []);
```
Comment (jclark) [inline compiler/modules/nback/init.bal]: Not using retValue.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/nback/build.bal [hunk 9]
```diff
@@ -393,3 +393,13 @@ function buildConstBoolean(boolean b) returns llvm:Value {
 function heapPointerType(llvm:Type ty) returns llvm:PointerType {
     return llvm:pointerType(ty, HEAP_ADDR_SPACE);
 }
+
+function buildFunctionSignature(bir:FunctionSignature signature) returns llvm:FunctionType|BuildError {
```
Comment (jclark) [inline compiler/modules/nback/build.bal]: This doesn't need to return BuildError any more does it?

File: compiler/compile.bal [hunk 10]
```diff
@@ -149,3 +154,21 @@ function subModuleSourceParts(string basename, bir:ModuleId id) returns front:So
 function subModuleSuffix(bir:ModuleId id) returns string {
     return ".".'join(...id.names.slice(1));
 }
+
+function buildInitModule(CompileContext cx, front:ResolvedModule entryMod) returns CompileError? {
```
Comment (jclark) [inline compiler/compile.bal]: We shouldn't use the same name for this as for the method on CompileContext when they do different things. buildModule is returning a llvm:MOdule, so I would call this something different, perhaps generateInitModule.

**Review by jclark — CHANGES_REQUESTED** [review]

**Pull request merged**